src/main/java/net/minecraftforge/server/permission/handler/IPermissionHandler.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.server.permission.handler;
+
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.util.StringRepresentable;
+import net.minecraftforge.server.permission.PermissionAPI;
+import net.minecraftforge.server.permission.events.PermissionGatherEvent;
+import net.minecraftforge.server.permission.nodes.PermissionDynamicContext;
+import net.minecraftforge.server.permission.nodes.PermissionNode;
+
+import java.util.Collection;
+
+/**
+ * This is the Heart of the PermissionAPI, it manages {@link PermissionNode}s
+ * as well as it handles all permission queries.
+ *
+ * <p>Note: You do not need to implement a PermissionHandler to query for permissions.</p>
+ *
+ * @implNote The {@link DefaultPermissionHandler} does forward all permission queries to the PermissionNodes default resolver.
+ * @apiNote You can implement your own PermissionHandler using the {@link PermissionGatherEvent.Handler} event.
+ */
+public interface IPermissionHandler
+{
+   /**
+    * {@return an identifier for the PermissionHandler, used for logging purposes}
+    */
+   ResourceLocation getPermissionHandlerName();
+
+   /**
+    * @param node to be registered, should not be called by mods.
+    */
+   void registerNode(PermissionNode<?> node);
+
+   /**
+    * {@return a collection of all registered Nodes}
+    */
+   Collection<PermissionNode<?>> getRegisteredNodes();
```

Comment:
Is this a modifiable or unmodifiable collection? Is this a snapshot of the registered nodes at the time the method was called, or is it a live view of all registered nodes?

Reply:
Would you say that specifying it as unmodifieable in javadoc is enough?
Or should I use the google.common (or whatever exactly the package is) ImmutableList?

Reply:
`Collections.unmodifiableCollection(original)`?

Reply:
Note that there's a difference between an immutable collection and an unmodifiable collection: the former means the list's contents will never change and cannot be changed through that object, while the latter only means the list's contents cannot be changed through that object -- it's possible for an unmodifiable collection to change its contents.

As @TheSilkMiner points out, using `Collections.unmodifiableCollection(Collection<T>)` creates an unmodifiable live view of the passed-in collection. So my recommendation is to mark the method as returning  `an unmodifiable view of the collection of registered permission nodes`, then return in the default permission handler a `Collections.unmodifiableCollection(registeredNodes)`. 

The return type shouldn't change (other than perhaps considering a move to `Set` instead of `Collection`, for the guarantee of there never being two `PermissionNode`s in the set which are `equals(Object)` with each other.